pkg/authconfigmap/mapidentity.go

```diff
@@ -0,0 +1,124 @@
+package authconfigmap
+
+import (
+	"encoding/json"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/pkg/errors"
+	"github.com/weaveworks/eksctl/pkg/iam"
+)
+
+// MapIdentity represents an IAM identity with an ARN.
+type MapIdentity struct {
+	iam.Identity `json:",inline"`
+	ARN          ARN `json:"-"` // This field is (un)marshaled manually
+}
+
+// UnmarshalJSON for MapIdentity makes sure there is either a "rolearn" or "userarn" key
+// and places the data of that key into the "ARN" field. All other fields are handled by
+// the default implementation of unmarshal.
+func (m *MapIdentity) UnmarshalJSON(data []byte) error {
+	// Handle everything as usual
+	type alias MapIdentity
+	var a alias
+	if err := json.Unmarshal(data, &a); err != nil {
+		return err
+	}
+
+	// We want to unmarshal "(rolearn|userarn)" into the "ARN" field and then unmarshal
+	// the rest as usual
+	var outerKeys map[string]*json.RawMessage
+	if err := json.Unmarshal(data, &outerKeys); err != nil {
+		return err
+	}
+
+	arnData, ok := outerKeys[roleKey]
+	if !ok {
+		arnData, ok = outerKeys[userKey]
+		if !ok {
+			return errors.New("missing arn")
+		}
+	}
+
+	var canonicalArn string
+	if err := json.Unmarshal(*arnData, &canonicalArn); err != nil {
+		return err
+	}
+
+	arn, err := arn.Parse(canonicalArn)
+	if err != nil {
+		return err
+	}
+	a.ARN = ARN(arn)
+
+	*m = MapIdentity(a)
+	return nil
+}
+
+// MarshalJSON for MapIdentity marshals the ARN field into either "rolearn" or "userarn"
+// depending on what is appropriate and returns and error if it cannot determine that.
+// All other fields are marshaled with the default marshaler
+func (m *MapIdentity) MarshalJSON() ([]byte, error) {
+	// Marshal everything as one would normally do
+	type alias MapIdentity
+	a := (*alias)(m)
```

Comment:
I'd just make a copy without creating a type alias, try: `json.Marshal(MapIdentity(*m))`.

Reply:
Ignore that, it doesn't solve the problem. I need to think of this a little more.

Reply:
What is the problem? It doesn't create a copy of `MapIdentity`, just `*MapIdentity`. Or at least that is the idea.

Reply:
Well, what you are trying to overcome is recursive call to `MarshalJSON`, right? I am not sure if what I suggested with `json.Marshal(MapIdentity(*m))` actually overcomes that or not, I'll need to test a few things and understand why this whole function is needed. I don't yet understand, but I'm keen to hear an explanation, if you have a time :)

Reply:
Yeah the point of the alias is to not have a recursive `MarshalJSON`. The current code works as expected, no?

So the problem is that `MapIdentity` either has the structure

```
{
   "rolearn": "..."
}
```

or

```
{
   "userarn": "..."
}
```

and in order to handle that situation in go you could have a `MapRole` and a `MapUser` struct and write custom unmarshaling logic that determines which type it should be. Since this is super annoying and the only time they need to be distinguished is when (un)marshaling I opted to create one type with custom functions instead. 

Since only the `ARN` field and the `rolearn` and `userarn` keys are relevant we can handle everything else with the default implementation of (un)marshal. In order  to do that we need to create a type alias since otherwise it will get stuck in infinite recursion.

Reply:
I think exposing two separate fields would simplify things. We will need to review the design of the struct as part of #874 anyway.
```Go
type Identity struct {
        RoleARN *ARN `json:"rolearn,omitempty"`
        UserARN *ARN `json:"userarn,omitempty"`
	Username *string   `json:"username,omitempty"`
	Groups   []string `json:"groups,omitempty"`
}
```

You could easily check whether `RoleARN` or `UserARN` is `nil` (having both set should probably result in an error).

It may make sense to add a higher-level struct, but let's not worry about it for now.

All in all, please do feel free to make changes to `pkg/iam/` as you see fit, there's no need to work around the API it currently exposes!

Reply:
Also, if you don't have the time - just let us know and we can take care of wrapping it up :)

Reply:
I'll take a look at it next week!

Reply:
I have now refactored to use such a struct.

Unfortunately there are new (and better I suppose) linters so I haven't been able to test it. I don't really have time to do the necessary rebase for this, sorry.

Apart from some question marks the functionality shouldn't really have changed though. The biggest question mark is in `pkg/ctl/create/iamidentitymapping.go:49` where an error should be handled. Perhaps it should be moved into the `doCreateIAMIdentityMapping` function?